ai2thor/server.py

```diff
@@ -22,6 +23,263 @@ def default(self, obj):
 
 
+class LazyInstanceDetections2D(Mapping):
+    def __init__(self, instance_masks):
+
+        self._detections2d = {}
+        self.instance_masks = instance_masks
+        self._loaded = False
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self.instance_masks == other.instance_masks
+        else:
+            return False
+
+    def mask_bounding_box(self, mask):
+        rows = np.any(mask, axis=1)
+        cols = np.any(mask, axis=0)
+        rw = np.where(rows)
+        if len(rw[0]) == 0:
+            return None
+
+        rmin, rmax = rw[0][[0, -1]]
+        cmin, cmax = np.where(cols)[0][[0, -1]]
+
+        return cmin, rmin, cmax, rmax
+
+    def __contains__(self, key):
+        return key in self.instance_masks
+
+    def __getitem__(self, key):
+        if key in self._detections2d:
+            return self._detections2d[key]
+
+        mask = self.instance_masks[key]
+        self._detections2d[key] = self.mask_bounding_box(mask)
+
+        return self._detections2d[key]
+
+    def __len__(self):
+        return len(self.instance_masks)
+
+    def __iter__(self):
+        for k in self.instance_masks.keys():
+            self.__getitem__(k)
+
+        return iter(self._detections2d)
+
+
+class LazyClassDetections2D(LazyInstanceDetections2D):
+
+    def __len__(self):
+        self._load_all()
+        return len(self._detections2d)
+
+    def _load_all(self):
+        if not self._loaded:
+            all_integer_keys = self.instance_masks.unique_integer_keys
+            for cls, colors in self.instance_masks.class_colors.items():
+                for color in colors:
+                    if self.instance_masks._integer_color_key(color) in all_integer_keys:
+                        self.__getitem__(cls)
+                        break
+
+        self._loaded = True
+
+    def __iter__(self):
+        self._load_all()
+
+        return iter(self._detections2d)
+
+    def __getitem__(self, cls):
+        if cls in self._detections2d:
+            return self._detections2d[cls]
+
+        detections = []
+
+        for color in self.instance_masks.class_colors.get(cls, []):
+            mask = self.instance_masks.instance_segmentation_frame_uint32 == self.instance_masks._integer_color_key(color)
+            bb = self.mask_bounding_box(mask)
+            if bb:
+                detections.append(bb)
+        if detections:
+            self._detections2d[cls] = detections
+        else:
+            raise KeyError(cls)
+
+        return self._detections2d[cls]
+
+class LazyMask(Mapping):
+
+    def __contains__(self, key):
+        return self.mask(key) is not None
+
+    def __getitem__(self, key):
+
+        m = self.mask(key)
+
+        if m is None:
+            raise KeyError(key)
+
+        return m
+
+    def __iter__(self):
+        self._load_all()
+        return iter(self._masks)
+
+    def __len__(self):
+        self._load_all()
```

Comment:
Should we make `_load_all` an `@abstractmethod` in `LazyMask`?

Reply:
good suggestion, done.